<div style="float: right; margin: 0px 15px 15px 0px;">
<img src="https://upload.wikimedia.org/wikipedia/commons/b/b6/HULT_IBS_Logo_Outline_Black_%28cropped%29.png" width=150/>
</div>
<h1> Python for Data Analysis: Methods & Tools </h1>
<em> <strong>Python for everyday people </strong></em>
<br><br>
Written by Felipe Dominguez - Professor Adjunct <br>
Hult International Business School <br>
<br>
<h1><u> Chapter 06 - Introduction to Pandas</u></h1>
<em>DataFrames structures</em>

<h3>1. Business Analytics & Python</h3><br>
<div align = "justify">
Yes! You know all the fundamentals of Python and you are ready to branch out into any number of specializations within Python. There are multiple programming fields that rely significantly in Python. Web development, Artificial Intelligence, Machine learning, Data Scientists, are just some examples of tasks you can accomplish by knowing Python. Furthermore, according to <a href = https://pypl.github.io/PYPL.html> PYPL (PopularitY of Programming Language)</a> Python, as of today, is still the most popular programming language worldwide, surpassing other languages like JavaScript, R, C/C++, etc.<br><br>
Now, as you know (hopefully), this course is related to <strong>Business Analytics</strong>. And, therefore, you will learn how to leverage Python to this end. Python is widely used for Data Science. One of the most important packages for it is the <strong>Pandas</strong> library. There are more libraries you will learn and work in the near future, like <strong>sci-kit learn</strong>, <strong>SciPy</strong>, or <strong>tensorflow</strong>. However, chapter 06 and 07, will focus specifically on learning how to start using and explore datasets with Pandas.<br><br>
But, what makes Python and Pandas so special? There are many reasons why using Pandas and Python it's better than Excel, but the most important are scalability, speed, automation, and interoperability.
</div>

~~~
i. Scalability - Python is limited by hardware.
ii. Speed - When working with large amount of data (thousands, millions), Python & Pandas are much faster than Excel.
iii. Automation - Automate tasks, even entire codes, reducing time spent on tedious and repetitive activities.
iv. Interoperability - You can easily connect Python with other data sources like SQL servers, to streamline   Data flows.
~~~
<br>
<strong>Okay, now let's dive into What is Pandas?</strong>




<h3>2. But...What is Pandas?</h3><br>
<div align = "justify">
According to the official documentation, <em><a href = "https://pandas.pydata.org/docs/">pandas</a> is an open source, BSD-licensed library providing high-performance, easy-to-use data structures and data analysis tools for the Python programming language.</em> In other words, Pandas is a Python library used for data manipulation and analysis. Furthermore, Pandas provides a set of functions and data structures that help you work with data and enables you to perform operations such as slicing, indexing, merging, and group by easily. <br><br>
The first thing you need to do, is to install the pandas library. To do so, open a new cell block and run:
</div>

~~~
!pip install pandas
~~~
<br>
<div align = "justify">
Once you install the package, you can remove the cell block from your code.<br><br>
<em><strong>Note:</strong> You need to use the <strong>exclamation mark (!)</strong> to run the code, if not Jupyter Notebook will not understand you are trying to install a new package. You must run this syntax (!pip install library) the first time you want to try a new library. After <strong>installing</strong> a library, you just need to <strong>import</strong> it at the beginning of each file.</em>
</div>



<h4>2.1. Pandas DataFrames</h4><br>
<div align = "justify">
One of the most important functionalities of Pandas package is that it gives you access to DataFrames structures. Pandas, DataFrames, what? Yes, it may not make any sense now, but I promise that you will get use to this language sooner than later. <br><br>
A DataFrame is a <strong>two-dimensional</strong> data structure used to store and manipulate data in a <strong>table format</strong>. DataFrames are key to create data analysis, machine learning models, or data visualization charts. Two-dimensional refers to a data structure which consists of a collection of <strong>rows and columns</strong>. Yes, exactly like any spreadsheet format! Each row represents an <stron>observation or a record of an event</stron> whereas each column represents a <strong>feature (property, characteristic, or attribute) of each observation</strong> (<em>in R columns are called variables</em>). Columns will be named from 0 to j if no header is provided whereas rows will be indexed from 0 to i. The following tabular structure represents a <strong>pandas DataFrame</strong> with 4 observations (rows) and 3 features (columns). <strong>See how the row index starts at 0</strong>.<br><br>
</div>

~~~
      name  age  income
0    Alice   25   50000
1      Bob   30   60000
2  Charlie   35   70000
3   Debbie   40   80000
        
~~~
<em><small>Table 1.1. DataFrame structure example</small></em><br><br>

<div align = "justify">
To use <strong>pandas</strong> you need to import the library into your code (if you already installed it). Just follow the syntax:
</div>

~~~
import library as library_abbreviation
~~~

<div align = "justify">
Even though you can use any word/name for the library abbreviation, there are some standards among the programming community for the most used packages. For example, numpy is always imported as np, seaborn as sns, or tensorflow as tf. In this case, pandas is <strong>always</strong> imported as <strong>pd</strong>. <br><br>Let's dive into the fantastic world of dataframes together.
     
    
</div>

In [3]:
# Code 2.1.
# import pandas library
import pandas as pd


<h4>2.2. My first DataFrame</h4><br>
During this chapter you will see there are multiple ways to create a DataFrame. You can create a DataFrame object based on dictionaries, lists, list of tuples. Furthermore, you can <strong>import structured data</strong> like Excel or csv,  files into a pandas.<br><br>
Let's start by creating Table 1.1. with pandas. Code 2.2. presents a a way to create a dataframe based on a dictionary whereas code 2.3. present the same dataframe build from a list of tuples.



In [3]:
# Code 2.2.


# create a dictionary based on 3 features
data = {'name': ['Alice', 'Bob', 'Charlie', 'Debbie'],
        'age': [25, 30, 35, 40],
        'income': [50000, 60000, 70000, 80000]}

# Create a dataframe from the dictionary
df = pd.DataFrame(data)

# Print the dataframe
print(df)


      name  age  income
0    Alice   25   50000
1      Bob   30   60000
2  Charlie   35   70000
3   Debbie   40   80000


In [26]:
# Code 2.3.
# No need of importing pandas again

# Create a list of tuples with 3 festures each
data_2 = [("Alice", 25, 50000), 
          ("Bob", 30, 60000),
          ("Charlie", 35, 70000),
          ("Debbie", 40, 80000)]

# Create a dataframe from the list of tuples
df_2 = pd.DataFrame(data_2, columns = ["name", "age", "income"] )

# Output dataframe (not print)
df_2

name  age  income
0    Alice   25   50000
1      Bob   30   60000
2  Charlie   35   70000
3   Debbie   40   80000

<div align = "justify">
<em><strong>Note: </strong>There is a small difference between Code 2.2. and 2.3. Did you catch it? On code 2.2. you are printing the DataFrame whereas in code 2.3. you are outputting it. As you can observed, there is a difference on how Python print and output a DataFrame. You can use any you want, just remember if you put any code line below output df_2 in code 2.3., Python won't print anything. If you explicitly wrap a DataFrame with the print function you are forcing Python for an output.</em>
    </div>

<h3>2. The history of the FIFA World Cup</h3><br>
Did you know that the first official World Cup was held in Uruguay in 1930, almost 92 years ago? A vibrant final was played between Argentina and Uruguay in front of a crowd of over 80,000 fans at Estadio Centenario in Montevideo. Uruguay emerged as the winners with a score of 4-2, making that the first time they had claimed the trophy. Since that first World Cup, 22 finals have been played (excluding 1942 and 1946), with a total of 2548 goals scored and 82 countries participating. <br><br>
In this exercise, you will use the pandas library to analyze a dataset containing all of the World Cup games since 1930. You will <strong>import the dataset into a Pandas DataFrame and learn how to work with it</strong>. <br><br>
Are you ready to get started? Let's dive in!<br><br>



<h4>2.1 Importing Excel/csv to Pandas</h4><br>
<div align = "justify">
One of the most useful features of the pandas library is its ability to import datasets from a variety of sources. In this section you will learn how to create a DataFrame based on an Excel or csv files using the functions <strong>read_excel</strong> and <strong>read_csv</strong>.<br><br>
Before diving directly into creating a DataFrame, let's read the Docstring of each function and understand what they are meant for and its parameters. Code 2.1.1. and 2.1.2. call the help function in each one.<br><br>
<h5>i. pd.read_csv()</h5><br>
The <strong>pd.read_csv()</strong> allows you to read any comma-separated document (csv) into a pandas DataFrame. It requires one mandatory argument (filepath_or_buffer) that tells Python where the file is located. There are many optional arguments that you can use to customize the data import. For example, the <strong>header</strong> argument define the row number to use as the column name and the start of the table. If no value is assigned, pandas assign the first row as the header by default.<br><br>

<h5>ii. pd.read_excel()</h5><br>
The <strong>pd.read_excel()</strong> allows you to read any excel document (xslx) into a pandas DataFrame. It requires one mandatory argument (io) which, similar pd.read_csv(), tells Python where the file is located. Unlike csv, Excel files can have multiple sheets. By default, pandas will read the first sheet of any Excel file. However, you can select a specific sheet by using the sheet_name optional argument.<br><br>
Look at Tables 2.1.1. and 2.1.2. to see some of the most common optional arguments for pd.read_excel() and pd.read_csv() functions, respectively.
    
    

<em><strong>Note: </strong>You can indicate the complete path of a document ("C:/Desktop/Documents/file_example.csv"). However, if the document is somewhere within the same folder as your jupyter file, you could use the "./" syntax that tells Python to start looking from the same folder as your file. For example, the world_cup_matches file is within the <strong>\__resources</strong> folder, so you can write the path as "./__resources/world_cup_matches.csv"</em>
</div>

In [14]:
# Code 2.1.1.
help(pd.read_csv)

Help on function read_csv in module pandas.io.parsers.readers:

read_csv(filepath_or_buffer: 'FilePath | ReadCsvBuffer[bytes] | ReadCsvBuffer[str]', sep=<no_default>, delimiter=None, header='infer', names=<no_default>, index_col=None, usecols=None, squeeze=None, prefix=<no_default>, mangle_dupe_cols=True, dtype: 'DtypeArg | None' = None, engine: 'CSVEngine | None' = None, converters=None, true_values=None, false_values=None, skipinitialspace=False, skiprows=None, skipfooter=0, nrows=None, na_values=None, keep_default_na=True, na_filter=True, verbose=False, skip_blank_lines=True, parse_dates=None, infer_datetime_format=False, keep_date_col=False, date_parser=None, dayfirst=False, cache_dates=True, iterator=False, chunksize=None, compression: 'CompressionOptions' = 'infer', thousands=None, decimal: 'str' = '.', lineterminator=None, quotechar='"', quoting=0, doublequote=True, escapechar=None, comment=None, encoding=None, encoding_errors: 'str | None' = 'strict', dialect=None, error_bad_li

In [15]:
# Code 2.1.2.
help(pd.read_excel)

Help on function read_excel in module pandas.io.excel._base:

read_excel(io, sheet_name: 'str | int | list[IntStrT] | None' = 0, header: 'int | Sequence[int] | None' = 0, names=None, index_col: 'int | Sequence[int] | None' = None, usecols=None, squeeze: 'bool | None' = None, dtype: 'DtypeArg | None' = None, engine: "Literal['xlrd', 'openpyxl', 'odf', 'pyxlsb'] | None" = None, converters=None, true_values: 'Iterable[Hashable] | None' = None, false_values: 'Iterable[Hashable] | None' = None, skiprows: 'Sequence[int] | int | Callable[[int], object] | None' = None, nrows: 'int | None' = None, na_values=None, keep_default_na: 'bool' = True, na_filter: 'bool' = True, verbose: 'bool' = False, parse_dates=False, date_parser=None, thousands: 'str | None' = None, decimal: 'str' = '.', comment: 'str | None' = None, skipfooter: 'int' = 0, convert_float: 'bool | None' = None, mangle_dupe_cols: 'bool' = True, storage_options: 'StorageOptions' = None) -> 'DataFrame | dict[IntStrT, DataFrame]'
    Rea

<div style = "width:image width px; font-size:80%; text-align:center;">
<table align="center">
<col width="80">
<col width="10">   
<col width="700">
    <tr>
        <th>Argument</th>
        <th>    | </th>
        <th>Description</th>
    </tr>
    <tr>
        <td>filepath_or_buffer</td>
        <td>    | </td>
        <td> the file, path, or URL of the data</td>
    </tr>
    <tr>
    <tr>
        <td>sep</td>
        <td>    | </td>
        <td>Delimiter to use, usually comma </td>
    </tr>
    <tr>
        <td>header</td>
        <td>    | </td>
        <td> Row number to use as the column names, and the start of the data.  Default behavior is to infer the column names</td>
    </tr>   
</table></div><br>
<div align = 'center'>
<em><small>Table 2.1.1.: Extremely useful arguments for pd_read_csv().</small></em>
</div>


<div style = "width:image width px; font-size:80%; text-align:center;">
<table align="center">
<col width="100">
<col width="10">   
<col width="600">
    <tr>
        <th>Argument</th>
        <th>    | </th>
        <th>Description</th>
    </tr>
    <tr>
        <td>io</td>
        <td>    | </td>
        <td> the file, path, or URL of the data</td>
    </tr>
    <tr>
    <tr>
        <td>sheet_name</td>
        <td>    | </td>
        <td> if your data exists on multiple sheets in Excel, this is the argument to tell Python which sheet to read</td>
    </tr>
    <tr>
        <td>header</td>
        <td>    | </td>
        <td> if your Excel file has column names in the first row, this is the argument to tell this to Python</td>
    </tr>   
    <tr>
        <td>dtype</td>
        <td>    | </td>
        <td> Excel and Python tend to interpret data types differently. This argument helps you control this.</td>
    </tr>
</table></div><br>
<div align = 'center'>
<em><small>Table 2.1.2.: Extremely useful arguments for pd_read_excel().</small></em>
</div>

It is a good practice to use a keyword to identify DataFrames objects. Usually, Data Scientists use the abbreviation <strong>df</strong> at the beginning or ending of a variable that will store a DataFrame.<br><br>
Code 2.1.3. shows how to import an Excel file whereas code 2.1.4. presents how to import a csv file. As you can observed, both codes produce the same outcome (because they are reading the same data, just different file types). Although the use of the optional arguments will not change the results of these codes (sheet_name, header, sep), it is a good practice to explicitly specify them, so you become aware they exist and can be used for other purposes. In the future, once you are fluent in Python, you may want to avoid them.

<em><strong>Note: </strong>See how Code 2.1.3 and 2.1.4. use the <strong>df_</strong> at the beginning of each DataFrame variable. 

In [15]:
# Code 2.1.3. Import Excel file

# storing path file
path_file = "./__resources/world_cup_matches.xlsx"

# transform excel file into DataFrame
df_wcup = pd.read_excel(io = path_file,  # path to file or URL
                        sheet_name = 0,  # or 'world_cup_matches' [optional argument]
                        header = 0)      # first row have the headers [optional argument]

print("df_wcup type is ", type(df_wcup), "\n")

# output DataFrame
df_wcup

df_wcup type is  <class 'pandas.core.frame.DataFrame'> 



ID  Year       Date           Stage      Home Team  Home Goals  \
0      1  1930 1930-07-13     Group stage         France           4   
1      2  1930 1930-07-13     Group stage  United States           3   
2      3  1930 1930-07-14     Group stage     Yugoslavia           2   
3      4  1930 1930-07-14     Group stage        Romania           3   
4      5  1930 1930-07-15     Group stage      Argentina           1   
..   ...   ...        ...             ...            ...         ...   
895  896  2018 2018-07-07  Quarter-finals         Russia           2   
896  897  2018 2018-07-10     Semi-finals         France           1   
897  898  2018 2018-07-11     Semi-finals        Croatia           2   
898  899  2018 2018-07-14     Third place        Belgium           2   
899  900  2018 2018-07-15           Final         France           4   

     Away Goals Away Team                    Win Conditions  Host Team  \
0             1    Mexico                               NaN      False   
1             0   Belgium                               NaN      False   
2             1    Brazil                               NaN      False   
3             1      Peru                               NaN      False   
4             0    France                               NaN      False   
..          ...       ...                               ...        ...   
895           2   Croatia  Croatia win on penalties (3 - 4)       True   
896           0   Belgium                               NaN      False   
897           1   England                        Extra time      False   
898           0   England                               NaN      False   
899           2   Croatia                               NaN      False   

            Winner  
0           France  
1    United States  
2       Yugoslavia  
3          Romania  
4        Argentina  
..             ...  
895        Croatia  
896         France  
897        Croatia  
898        Belgium  
899         France  

[900 rows x 11 columns]

In [13]:
# Code 2.1.4. Import csv file
# storing path file
path_file = "./__resources/world_cup_matches.csv"

# transform csv file into DataFrame
df_wcup_2 = pd.read_csv(filepath_or_buffer = path_file, # path to file or URL
                        sep = ",",                      # separator character
                        header = 0)                     # first row headers

print("df_wcup_2 type is ", type(df_wcup_2), "\n")

# output DataFrame
df_wcup_2

df_wcup_2 type is  <class 'pandas.core.frame.DataFrame'> 



ID  Year     Date           Stage      Home Team  Home Goals  \
0      1  1930  7/13/30     Group stage         France           4   
1      2  1930  7/13/30     Group stage  United States           3   
2      3  1930  7/14/30     Group stage     Yugoslavia           2   
3      4  1930  7/14/30     Group stage        Romania           3   
4      5  1930  7/15/30     Group stage      Argentina           1   
..   ...   ...      ...             ...            ...         ...   
895  896  2018   7/7/18  Quarter-finals         Russia           2   
896  897  2018  7/10/18     Semi-finals         France           1   
897  898  2018  7/11/18     Semi-finals        Croatia           2   
898  899  2018  7/14/18     Third place        Belgium           2   
899  900  2018  7/15/18           Final         France           4   

     Away Goals Away Team                    Win Conditions  Host Team  \
0             1    Mexico                               NaN      False   
1             0   Belgium                               NaN      False   
2             1    Brazil                               NaN      False   
3             1      Peru                               NaN      False   
4             0    France                               NaN      False   
..          ...       ...                               ...        ...   
895           2   Croatia  Croatia win on penalties (3 - 4)       True   
896           0   Belgium                               NaN      False   
897           1   England                        Extra time      False   
898           0   England                               NaN      False   
899           2   Croatia                               NaN      False   

            Winner  
0           France  
1    United States  
2       Yugoslavia  
3          Romania  
4        Argentina  
..             ...  
895        Croatia  
896         France  
897        Croatia  
898        Belgium  
899         France  

[900 rows x 11 columns]

<h4>2.2. First thing to do!</h4><br><br>
<div align = "center">
<strong>View an entire dataset in Python is a bad idea</strong>
</div><br>
<div justify = "align">
As you can see on Code 2.1.3. and 2.1.4., Python is outputting the full world cup table. Even though this is correct, it is not a good practice. Having the full table makes your code messy and less readable. Doing so will be the equivalent of printing in paper a full Excel worksheet with thousands of rows. Additionally, you don't need to see all the observations in a DataFrame, you just need to have a high-overview of what this DataFrame contains. <br><br>There are two pandas functions that will help you to get a better overview of a DataFrame.<br> </div>

~~~
i. pd.head()
ii. pd.info()
~~~
<div align = "justify">
<h5>i. pd.DataFrame.head()</h5><br>
The <strong>pd.DataFrame.head()</strong> method allows you to control how many rows you want to display in the output. head() by default will display the first 5 rows of any DataFrame. However, you can customize the number of row by passing the n argument with the numbers of rows you want. <br><br>
But, why the head function is so useful and why it should be the first thing to do when importing a DataFrame? Well, running the head function (and getting the expected outcome) confirm you imported a DataFrame correctly. Additionally, you will get a brief glance at the quality of the data and its features (columns).
    
<em><strong>Note: </strong>If you want to see the last observations of a DataFrame, you can use the function <strong>tail()</strong>.</em>

<h5>ii. pd.DataFrame.info()</h5><br>
The <strong>pd.DataFrame.info()</strong> prints information about a DataFrame, including the column data type, label, non-null values and memory usage. This method is the first way to evaluate the quality of your data. You can see how many nulls values are present in your DataFrame. In future classes you will learn how to deal and handle null values. <br><br>
As an example, in code 2.2.2. you can easily check that the df_wcup DataFrame has 900 observations, the column "Win Conditions" has mostly null values, and you have datetime, boolean, numeric, and string columns.
    
<em><strong>Note: </strong>row-wise elements can have different data types. However, data types must be the same for each column element. If Python found different data types elements within the same column, it will transform the column fully into a string (object) even though there was only item different. <br><br>Look at table 2.2.</em>
</div>


In [16]:
# Code 2.2.1.

# Change n to whatever number you like.
df_wcup.head(n = 5)

ID  Year       Date        Stage      Home Team  Home Goals  Away Goals  \
0   1  1930 1930-07-13  Group stage         France           4           1   
1   2  1930 1930-07-13  Group stage  United States           3           0   
2   3  1930 1930-07-14  Group stage     Yugoslavia           2           1   
3   4  1930 1930-07-14  Group stage        Romania           3           1   
4   5  1930 1930-07-15  Group stage      Argentina           1           0   

  Away Team Win Conditions  Host Team         Winner  
0    Mexico            NaN      False         France  
1   Belgium            NaN      False  United States  
2    Brazil            NaN      False     Yugoslavia  
3      Peru            NaN      False        Romania  
4    France            NaN      False      Argentina

In [41]:
# Code 2.2.2.
df_wcup.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 900 entries, 0 to 899
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   ID              900 non-null    int64         
 1   Year            900 non-null    int64         
 2   Date            900 non-null    datetime64[ns]
 3   Stage           900 non-null    object        
 4   Home Team       900 non-null    object        
 5   Home Goals      900 non-null    int64         
 6   Away Goals      900 non-null    int64         
 7   Away Team       900 non-null    object        
 8   Win Conditions  62 non-null     object        
 9   Host Team       900 non-null    bool          
 10  Winner          900 non-null    object        
dtypes: bool(1), datetime64[ns](1), int64(4), object(5)
memory usage: 71.3+ KB


<table align="center">
<col width="20">
<col width="10"> 
<col width="20">
<col width="10">  
<col width="30">
<col width="10">   
<col width="80">
<col width="10">  
<col width="40">
<col width="10">  
<col width="80">
<col width="10">  
    <tr>
        <th style="text-align: center">index</th>
        <th>    | </th>
        <th style="text-align: center">ID</th>
        <th>    | </th>
        <th style="text-align: left">Year</th>
        <th>    | </th>
        <th style="text-align: center">Date</th>
        <th>    | </th>
        <th style="text-align: left">Stage</th>
        <th>    | </th>
        <th style="text-align: left">Home Team</th>
    </tr>
    <tr>
        <td style="text-align: center">0</td>
        <td>    | </td>
        <td style="text-align: center">'int'</td>
        <td>    | </td>
        <td style="text-align: left">'int'</td>
        <td>    | </td>
        <td style="text-align: left">'datetime64'</td>
        <td>    | </td>
        <td style="text-align: left">Object</td>
        <td>    | </td>
        <td style="text-align: left">Object</td>
    </tr>
    <tr>
        <td style="text-align: center">1</td>
        <td>    | </td>
        <td style="text-align: center">'int'</td>
        <td>    | </td>
        <td style="text-align: left">'int'</td>
        <td>    | </td>
        <td style="text-align: left">'datetime64'</td>
        <td>    | </td>
        <td style="text-align: left">Object</td>
        <td>    | </td>
        <td style="text-align: left">Object</td>
    </tr>
    <tr>
        <td style="text-align: center">2</td>
        <td>    | </td>
        <td style="text-align: center">'int'</td>
        <td>    | </td>
        <td style="text-align: left">'int'</td>
        <td>    | </td>
        <td style="text-align: left">'datetime64'</td>
        <td>    | </td>
        <td style="text-align: left">Object</td>
        <td>    | </td>
        <td style="text-align: left">Object</td>
    </tr>   
    <tr>
        <td style="text-align: center">3</td>
        <td>    | </td>
        <td style="text-align: center">'int'</td>
        <td>    | </td>
        <td style="text-align: left">'int'</td>
        <td>    | </td>
        <td style="text-align: left">'datetime64'</td>
        <td>    | </td>
        <td style="text-align: left">Object</td>
        <td>    | </td>
        <td style="text-align: left">Object</td>
    </tr>
    <tr>
        <td style="text-align: center">4</td>
        <td>    | </td>
        <td style="text-align: center">'int'</td>
        <td>    | </td>
        <td style="text-align: left">'int'</td>
        <td>    | </td>
        <td style="text-align: left">'datetime64'</td>
        <td>    | </td>
        <td style="text-align: left">Object</td>
        <td>    | </td>
        <td style="text-align: left">Object</td>
    </tr>
</table>
<div align = 'center'>
<em>Table 2.3. Data types of df_wcup</em>
</div>  

<br><br><h4>2.3. The DataFrame structure</h4><br>
<div align = 'justify'>
<h5>2.3.1. Series Objects</h5><br>
As mentioned before, Pandas DataFrames can be thought as a Python spreadsheet compose by rows and columns. Another way to think about DataFrames is as a <strong>collection of two or more Series objects with the same index</strong>. But what are Series?<br><br>
A <strong>pandas series</strong> is a one-dimensional data structure (only 1 column) composed by a key-value pair (like dictionaries), where the keys are the index and the values are the values stores on that index. To create a pandas series object you can use the function <strong>pd.Series()</strong> as code 2.3.1.1. present.<br><br>
Each column of a pandas DataFrame can be considered as a pandas Series object. Therefore, you can access each column by calling the name of a serie. For example, to access one specific column, you can call the DataFrame object with the label of the column. In code 2.3.1.2. the column "Winner" was subset from the df_wcup DataFrame and store in a the winner_serie variable. <br><br>
The syntax to subset a DataFrame series is as below.
</div>

~~~
DataFrame['column name']
~~~
<div align = 'justify'>
As you can observed on code 2.3.1.2., pandas Series do not have the same output neither can be treated as pandas DataFrame. You can continue keeping the DataFrame type by adding <strong>[ ]</strong> to the column name as it were a list.
Additionally, <strong>you can subset more than one column from a DataFrame</strong>. In this case, you must use the square brackets to call the different columns and store a new DataFrame object. <br><br>
</div>

    
~~~
- Subset 1 column
DataFrame[ ['column name'] ]

- Subset 2 or more columns
DataFrame[['c_name_1', 'c_name_2',..., 'c_name_n']]
~~~
<div align = 'justify'>
<br>
Code 2.3.1.4. subset the columns Winner, Date, and Home Team from the df_wcup object. As you can observed, the order of the column names is not important.
</div>



In [61]:
# Code 2.3.1.1.
# Create a series object
my_series = pd.Series(("first", "second", "third"))
print(my_series)

0     first
1    second
2     third
dtype: object

In [88]:
# Code 2.3.1.2.
winner_serie = df_wcup['Winner']
print(type(winner_serie))
print(winner_serie)

<class 'pandas.core.series.Series'>
0             France
1      United States
2         Yugoslavia
3            Romania
4          Argentina
           ...      
895          Croatia
896           France
897          Croatia
898          Belgium
899           France
Name: Winner, Length: 900, dtype: object


In [70]:
# Code 2.3.1.3.
winner_df = df_wcup[['Winner']]
print(type(winner_df))
winner_df.head()

<class 'pandas.core.frame.DataFrame'>


Winner
0         France
1  United States
2     Yugoslavia
3        Romania
4      Argentina

In [89]:
# Code 2.3.1.4.
winner_df = df_wcup[['Winner', 'Date', 'Home Team']]
print(type(winner_df))
winner_df.head()

<class 'pandas.core.frame.DataFrame'>


Winner       Date      Home Team
0         France 1930-07-13         France
1  United States 1930-07-13  United States
2     Yugoslavia 1930-07-14     Yugoslavia
3        Romania 1930-07-14        Romania
4      Argentina 1930-07-15      Argentina

In [106]:
# Code 2.3.1.5.

# attribute call: DataFrame.column_name
wcup_winner = df_wcup.Winner
print(type(wcup_winner))
wcup_winner

<class 'pandas.core.series.Series'>


0             France
1      United States
2         Yugoslavia
3            Romania
4          Argentina
           ...      
895          Croatia
896           France
897          Croatia
898          Belgium
899           France
Name: Winner, Length: 900, dtype: object

<h3>3. Accessing DataFrames</h3><br>
<div align = 'justify'>
There are multiple ways to access the data within a DataFrame. As explain before, you can subset a DataFrame by typing the column (series) name. This method is called <strong>subset</strong>. Additionally, you can access the data by indexing or slicing a DataFrame. Let's dive into each of these methods.
</div>

<h4>3.1. Slicing DataFrames by rows</h4><br>
<div align = "justify">
It is possible to slice pandas Series and DataFrames by rows using square brackets ([]) and indicating how many rows you want to slice. Even though this is correct, it is highly recommended to use the <strong>.iloc</strong> mehtod that will be covered later.<br><br>
For now, let's understand the syntax of basic slicing.
</div>

~~~
DataFrame[start_row : end_row]
~~~
<div align = "justify">
As observed in code 3.1., the code df_wcup[5:10] returns the rows from the <strong>index 5 to the index 9</strong>.
    
<em><strong>Note: </strong>Do not confuse index with the ID column. Pandas will <strong>always</strong> assign an index number, starting by 0, to each observation (row). In this case, the ID column was given with the dataset. However, it's not related at all with the index.
</div>

In [86]:
# Code 3.1.
# Slicing the elements from 5 to 9 index.
sliced_df = df_wcup[5:10]
sliced_df

ID  Year       Date        Stage      Home Team  Home Goals  Away Goals  \
5   6  1930 1930-07-16  Group stage          Chile           3           0   
6   7  1930 1930-07-17  Group stage     Yugoslavia           4           0   
7   8  1930 1930-07-17  Group stage  United States           3           0   
8   9  1930 1930-07-18  Group stage        Uruguay           1           0   
9  10  1930 1930-07-19  Group stage          Chile           1           0   

  Away Team Win Conditions  Host Team         Winner  
5    Mexico            NaN      False          Chile  
6   Bolivia            NaN      False     Yugoslavia  
7  Paraguay            NaN      False  United States  
8      Peru            NaN       True        Uruguay  
9    France            NaN      False          Chile

<h4>3.2.  Subsetting DataFrames</h4><br>
<div align = 'justify'>
Amazing! You know how to slice DataFrames based on the index and how to subset a DataFrame based on its colum's names. But, what if you want to select a specific number of rows and specific columns? Certainly, you can combine both methods as code 3.2. presents. Even though this might seem correct, there are some limitations, specially considering pandas does not know what the type of data is contain in a DataFrame. For this reason, for production code it's always (highly / must) recommended to use the <strong>.iloc()</strong> and <strong>loc()</strong> pandas optimized access methods. Both methods (iloc and loc) accomplish the same results. However, one of them might fit better for certain situations you will encounter in the future.

<em><strong>Note: </strong><a href = "https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html">Look at the documentation about indexing and selecting data of pandas</a></em>
</div>


In [93]:
# Code 3.2.
# Not recommended method to subset and slice a DataFrame
df_wcup[['Date', 'Winner']][:2]

Date         Winner
0 1930-07-13         France
1 1930-07-13  United States

<h5>3.2.1. The .loc() method</h5><br>
<div align = 'justify'>
The <strong>.loc()</strong> method allows you to subset a DataFrame based on <strong>row index</strong> and <strong>column names</strong>. DataFrames are mutable, so positions of columns can change. Using the column names allows to ensure results are consistent in case a the DataFrame change its shape or order. Additionally, by using labels you are giving yourself (and others) a cognitive shortcut. It will be easier to remember what you were trying to do if you have df_wcup[ : , 'Winner'] rather than df_wcup[:, 10].
</div>

~~~
DataFrame.loc[ row NUMBERS, column NAMES]
~~~

In code 3.2.1.1. you can see how to use the .loc() methods to subset 1 or more columns and rows.

In [5]:
# Code 3.2.1.1.
# Use of .loc()
# Subset 10 rows and the column Winner, Date, and Home Team from df_wcup
df_loc = df_wcup.loc[ 20:30 , ["Winner", "Date", "Home Team"]] 
df_loc

Winner       Date       Home Team
20     Switzerland 1934-05-27     Switzerland
21          Sweden 1934-05-27          Sweden
22         Germany 1934-05-27         Germany
23           Spain 1934-05-27           Spain
24           Italy 1934-05-27           Italy
25  Czechoslovakia 1934-05-27  Czechoslovakia
26  Czechoslovakia 1934-05-31  Czechoslovakia
27         Germany 1934-05-31         Germany
28           Spain 1934-05-31           Italy
29         Austria 1934-05-31         Austria
30           Italy 1934-06-01           Italy

<h5>3.2.2. The .iloc() method</h5>
<div align = 'justify'>
The <strong>.iloc()</strong> method allows you to subset a DataFrame based on <strong>row index</strong> and <strong>column index</strong>. When getting data from different places, column names may not necessarily tell you what they are, you may no have the name of the columns, or the labels might be too large or complicated. In this cases, subsetting a DataFrame based on column index makes more sense than by the level. Additionally, using number instead of labels is more efficient and easier to loop.
</div>

~~~
DataFrame.loc[ row NUMBERS, column NUMBERS]
~~~

In code 3.2.2.2. you can see how to use the .iloc() methods to subset 1 or more columns and rows.

<em><strong>Note: </strong>Look that the iloc method exclude the end point of the range whereas loc method includes the end point range</em>

In [104]:
# Code 3.2.2.2.
# Subsetting with iloc() method
# Subset 10 rows and the column Winner, Date, and Home Team from df_wcup
df_iloc = df_wcup.iloc[ 20:30 , [10, 2, 4]] 
df_iloc

Winner       Date       Home Team
20     Switzerland 1934-05-27     Switzerland
21          Sweden 1934-05-27          Sweden
22         Germany 1934-05-27         Germany
23           Spain 1934-05-27           Spain
24           Italy 1934-05-27           Italy
25  Czechoslovakia 1934-05-27  Czechoslovakia
26  Czechoslovakia 1934-05-31  Czechoslovakia
27         Germany 1934-05-31         Germany
28           Spain 1934-05-31           Italy
29         Austria 1934-05-31         Austria

<h4>3.3. Slicing with conditions </h4><br>
<div align = 'justify'>
<strong>Conditional subsetting</strong> (or filter) can be accomplished using <strong>square brackets []</strong>.
There are multiple ways to slice a Data Frame based on multiple conditions. In this chapter you will learn how to slice using the .loc() function and using only square brackets.
</div>

~~~
i. DataFrame.loc[ rows, columns ][ condition 1 ][ condition 2 ][ condition n ]
ii. DataFrame[ ([condition_1 ]) & (condition_2) & ... & [condition_n] ]
~~~

<div align = 'justify'>
<h5>i. Conditions & loc()</h5><br>
Let's identify how many times Brazil has won the FIFA World Cup. In order to do so, you will need to filter the DataFrame where the Winner feature (column) is equal to Brazil. Code 3.3.1. present the method to slice the World Cup DataFrame based on this condition. As you can see, the semicolon is used to indicate which rows and columns you want to apply the slice.<br><br>
After the .loc() method is call, you can indicate the boolean condition. Pandas will look into each row of the DataFrame and will keep only those rows where the condition is True.<br><br>
Code 3.3.1. is not enough to identify how many times Brazil has been the champion. Right now, the result provides all the games where Brazil has won, including all World Cup stages. Therefore, you need to add a new condition to tell pandas to only consider those games in the Final stage and were Brazil won. Code 3.3.2. provides the answer to this question.
    
</div>


In [24]:
# Code 3.3.1.

# Condition to slice the world cup df
df_brazil_w = df_wcup.loc[:, :]             \
             [df_wcup['Winner'] == "Brazil"]

df_brazil_w.head()

ID  Year       Date           Stage Home Team  Home Goals  Away Goals  \
11  12  1930 1930-07-20     Group stage    Brazil           4           0   
40  41  1938 1938-06-05     Round of 16    Brazil           6           5   
48  49  1938 1938-06-14  Quarter-finals    Brazil           2           1   
51  52  1938 1938-06-19     Third place    Brazil           4           2   
53  54  1950 1950-06-24     First round    Brazil           4           0   

         Away Team Win Conditions  Host Team  Winner  
11         Bolivia            NaN      False  Brazil  
40          Poland     Extra time      False  Brazil  
48  Czechoslovakia            NaN      False  Brazil  
51          Sweden            NaN      False  Brazil  
53          Mexico            NaN       True  Brazil

In [25]:
# Code 3.3.2.
df_brazil_w = df_wcup.loc[:, :]               \
              [df_wcup['Winner'] == "Brazil"] \
              [df_wcup['Stage'] == "Final"]
df_brazil_w


/var/folders/1k/brqb_p7s34s2ycd6vzfrbxf80000gn/T/ipykernel_25440/1400912836.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  df_brazil_w = df_wcup.loc[:, :]               \


ID  Year       Date  Stage Home Team  Home Goals  Away Goals  \
135  136  1958 1958-06-29  Final    Sweden           2           5   
167  168  1962 1962-06-17  Final    Brazil           3           1   
231  232  1970 1970-06-21  Final    Brazil           4           1   
515  516  1994 1994-07-17  Final    Brazil           0           0   
643  644  2002 2002-06-30  Final   Germany           0           2   

          Away Team                   Win Conditions  Host Team  Winner  
135          Brazil                              NaN       True  Brazil  
167  Czechoslovakia                              NaN      False  Brazil  
231           Italy                              NaN      False  Brazil  
515           Italy  Brazil win on penalties (3 - 2)      False  Brazil  
643          Brazil                              NaN      False  Brazil

<h5>ii. Slice conditions</h5><br>
The second way to slice a DataFrame based on conditions, is to write the conditions within the square brackets directly. When you use the .loc() you have the ability to control where do you want the slice to apply. For example, you can slice over the complete DataFrame ([:, :]) or you could decide to slice based on a subset of the DataFrame ([10:20, :]). <br><br>
Let's look at code 3.3.3. and see how to slice a DataFrame directly based on boolean conditions to get the same result as code 3.3.2.

In [130]:
# Code 3.3.3.

df_brazil_w = df_wcup[(df_wcup['Winner'] == "Brazil") & (df_wcup['Stage'] == "Final")]
df_brazil_w

ID  Year       Date  Stage Home Team  Home Goals  Away Goals  \
135  136  1958 1958-06-29  Final    Sweden           2           5   
167  168  1962 1962-06-17  Final    Brazil           3           1   
231  232  1970 1970-06-21  Final    Brazil           4           1   
515  516  1994 1994-07-17  Final    Brazil           0           0   
643  644  2002 2002-06-30  Final   Germany           0           2   

          Away Team                   Win Conditions  Host Team  Winner  
135          Brazil                              NaN       True  Brazil  
167  Czechoslovakia                              NaN      False  Brazil  
231           Italy                              NaN      False  Brazil  
515           Italy  Brazil win on penalties (3 - 2)      False  Brazil  
643          Brazil                              NaN      False  Brazil

<h4>3.4. Setting values</h4><br>
<div align = 'justify'>
Updating values or creating new columns in a pandas DataFrame is very simple. You just need to assign the new value in the observation (or collection of observations) you want to update. It is highly-recommended to use a copy of the original dataset, so you can keep the original values immutable. The 'Win Conditions' column in the df_wcup DataFrame, only has a value if the game went to extra time or penalties. Let's update the null values (NaN), so it's says the Win Conditions was 'Full Time' for the rest of the games. <br><br>
Code 3.4.2. 

Similarly, you can create a new column in your DataFrame. You can create a new column by calling a column name that didn't exist before in the DataFrame. Once you assign the value, Pandas will understand you want create a new value and will match the index of the of each observation with the new column.<br><br>
Code 3.4.2. creates a new column named 'Total Goals', that is just the addition of Home and Away goals.
</div>

~~~
DataFrame['new_column'] =  new_value
~~~

<em><strong>Note 1: </strong>Any change you do to a DataFrame within Python will not affect the original data set unless it's exported. </em><br>
<em><strong>Note 2: </strong>Code 3.4.2. does not have the 'Full Time' that was replaced in code 3.4.1. That's because code 3.4.1. didn't replace the original dataset. </em>

In [186]:
# Code 3.4.1.
# Copy the original df
df_b_copy = df_brazil_w.copy()
# replacing the value only on those observations where 'Win Conditions' is null (NaN)
df_b_copy['Win Conditions'][df_b_copy['Win Conditions'].isnull()] = 'Full Time'
df_b_copy

/var/folders/1k/brqb_p7s34s2ycd6vzfrbxf80000gn/T/ipykernel_50074/3413299687.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_copy['Win Conditions'][df_copy['Win Conditions'].isnull()] = 'Full Time'


ID  Year       Date  Stage Home Team  Home Goals  Away Goals  \
135  136  1958 1958-06-29  Final    Sweden           2           5   
167  168  1962 1962-06-17  Final    Brazil           3           1   
231  232  1970 1970-06-21  Final    Brazil           4           1   
515  516  1994 1994-07-17  Final    Brazil           0           0   
643  644  2002 2002-06-30  Final   Germany           0           2   

          Away Team                   Win Conditions  Host Team  Winner  
135          Brazil                        Full Time       True  Brazil  
167  Czechoslovakia                        Full Time      False  Brazil  
231           Italy                        Full Time      False  Brazil  
515           Italy  Brazil win on penalties (3 - 2)      False  Brazil  
643          Brazil                        Full Time      False  Brazil

In [31]:
# Code 3.4.1.
# Copy the original df
df_b_copy = df_brazil_w.copy()
# replacing the value only on those observations where 'Win Conditions' is null (NaN) (using .loc)
df_b_copy.loc[df_b_copy['Win Conditions'].isnull(), ["Win Conditions"]] = 'Full Time'
df_b_copy

      ID  Year       Date  Stage Home Team  Home Goals  Away Goals  \
135  136  1958 1958-06-29  Final    Sweden           2           5   
167  168  1962 1962-06-17  Final    Brazil           3           1   
231  232  1970 1970-06-21  Final    Brazil           4           1   
515  516  1994 1994-07-17  Final    Brazil           0           0   
643  644  2002 2002-06-30  Final   Germany           0           2   

          Away Team                   Win Conditions  Host Team  Winner  
135          Brazil                              NaN       True  Brazil  
167  Czechoslovakia                              NaN      False  Brazil  
231           Italy                              NaN      False  Brazil  
515           Italy  Brazil win on penalties (3 - 2)      False  Brazil  
643          Brazil                              NaN      False  Brazil  


ID  Year       Date  Stage Home Team  Home Goals  Away Goals  \
135  136  1958 1958-06-29  Final    Sweden           2           5   
167  168  1962 1962-06-17  Final    Brazil           3           1   
231  232  1970 1970-06-21  Final    Brazil           4           1   
515  516  1994 1994-07-17  Final    Brazil           0           0   
643  644  2002 2002-06-30  Final   Germany           0           2   

          Away Team                   Win Conditions  Host Team  Winner  
135          Brazil                        Full Time       True  Brazil  
167  Czechoslovakia                        Full Time      False  Brazil  
231           Italy                        Full Time      False  Brazil  
515           Italy  Brazil win on penalties (3 - 2)      False  Brazil  
643          Brazil                        Full Time      False  Brazil

In [187]:
# Code 3.4.2.
# Copy the original df
df_b_copy_2 = df_brazil_w.copy()

# Creating total goals column as the sum between Home Goals & Away Goals
df_b_copy_2['Total Goals'] = df_b_copy_2['Home Goals'] + df_b_copy_2['Away Goals']
df_b_copy_2

ID  Year       Date  Stage Home Team  Home Goals  Away Goals  \
135  136  1958 1958-06-29  Final    Sweden           2           5   
167  168  1962 1962-06-17  Final    Brazil           3           1   
231  232  1970 1970-06-21  Final    Brazil           4           1   
515  516  1994 1994-07-17  Final    Brazil           0           0   
643  644  2002 2002-06-30  Final   Germany           0           2   

          Away Team                   Win Conditions  Host Team  Winner  \
135          Brazil                              NaN       True  Brazil   
167  Czechoslovakia                              NaN      False  Brazil   
231           Italy                              NaN      False  Brazil   
515           Italy  Brazil win on penalties (3 - 2)      False  Brazil   
643          Brazil                              NaN      False  Brazil   

     Total Goals  
135            7  
167            4  
231            5  
515            0  
643            2

<h4>3.5. Export a DataFrame</h4><br>
DataFrames can also be exported out of Pythons environment. Similar to import a file into pandas, you can export a DataFrame to multiple data sources. The methods <strong>to_excel( )</strong> and <strong>to_csv( )</strong> have several optional arguments that you can find in its documentations. When you export a DataFrame to excel or csv, pandas will automatically create a new column containing the dataset's index values (135, 167, 231, 515, 643), even if an index already exists. For now, this behavior might not affect your process or seems not important. However, it is extremely valuable in situations when the data has been divided into various parts as it supports to keep track of its original indexes. To remove this column, just set the optional argument <strong>index</strong> to False.

Code 3.5.1 and 3.5.2. shows how to export csv and excel files. 

In [192]:
df_brazil_w.to_csv("./brazil_wins.csv",
                   index = False)

In [193]:
df_brazil_w.to_excel("./brazil_wins.xlsx")

<h3>4. Summary</h3><br>
Congrats! You've learned the basics of Pandas!
Get ready to dive deeper into this library in your near future!


DataSource<br>
Maven Analytics<br>
https://www.mavenanalytics.io/data-playground

